### PR TITLE
refactor: disable prediction with ccb and cb_type mtr

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/cb_adf.cc [hunk 1]
```diff
@@ -261,6 +263,11 @@ void cb_adf::learn_MTR(multi_learner& base, multi_ex& examples)
   examples[_gen_cs.mtr_example]->num_features = nf;
   examples[_gen_cs.mtr_example]->weight = old_weight;
   std::swap(_gen_cs.mtr_ec_seq[0]->pred.a_s, _a_s_mtr_cs);
+  if (_ccb)
```
Comment (jackgerrits) [inline vowpalwabbit/cb_adf.cc:266]: It is illegal for this reduction to set a label other than its specified output label type.

File: vowpalwabbit/cb_adf.cc [hunk 2]
```diff
@@ -485,6 +492,7 @@ base_learner* cb_adf_setup(VW::setup_base_i& stack_builder)
   bool rank_all;
   float clip_p;
   bool no_predict;
+  bool ccb = false;
```
Comment (jackgerrits) [inline vowpalwabbit/cb_adf.cc:495]: This reduction knowing about ccb is a big break of encapsulation.

Comment (olgavrou) [reply vowpalwabbit/cb_adf.cc:495]: Would it make sense to use the same condition as `conditional_contextual_bandit` setup, i.e. `if (options.was_supplied("quiet") && !options.was_supplied("p")) { data->no_pred = true; }` and have a no-pred boolean here too?

File: vowpalwabbit/cb_adf.cc [hunk 3]
```diff
@@ -261,6 +263,11 @@ void cb_adf::learn_MTR(multi_learner& base, multi_ex& examples)
   examples[_gen_cs.mtr_example]->num_features = nf;
   examples[_gen_cs.mtr_example]->weight = old_weight;
   std::swap(_gen_cs.mtr_ec_seq[0]->pred.a_s, _a_s_mtr_cs);
+  if (_ccb)
+  {
+    examples[0]->l.conditional_contextual_bandit.explicit_included_actions.clear();
```
Comment (olgavrou) [inline vowpalwabbit/cb_adf.cc:268]: Could re-use `reduction_features` to communicate extra information between reductions that don't fit in the label or prediction and explicitly extend `ccb_reduction_features.h` to hold a field specifically for this

**Review by olgavrou — COMMENTED** [review]
Can we add a test to exercise this code?

File: vowpalwabbit/cb_adf.cc (hunk 3 above)
Comment (jackgerrits) [reply vowpalwabbit/cb_adf.cc:268]: I would say that modifying the reduction features of a problem type other than yourself is also a break of label/prediction layering rules

Comment (olgavrou) [reply vowpalwabbit/cb_adf.cc:268]: hm so then the only way is via the `pred.a_s` but somehow making it clear that it's not a prediction?

Comment (jackgerrits) [reply vowpalwabbit/cb_adf.cc:268]: Or by returned a different prediction type when this mode is enabled

File: vowpalwabbit/cb_adf.cc (hunk 1 above)
Comment (bassmang) [reply vowpalwabbit/cb_adf.cc]: Yes I just had this here to see the perf impact, I removed all changes to cb_adf in this PR

File: vowpalwabbit/cb_adf.cc (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/cb_adf.cc]: It can all be handled in ccb so I removed this as well

File: vowpalwabbit/cb_adf.cc (hunk 3 above)
Comment (bassmang) [reply vowpalwabbit/cb_adf.cc]: This step isn't necessary and actually messed with some other CCB functionality. The cb labels and predictions are already in ccb so we dont need to do anything here

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 4]
```diff
@@ -211,6 +212,26 @@ void save_action_scores(ccb& data, decision_scores_t& decision_scores)
   data.shared->pred.a_s.clear();
 }
 
+void remove_labeled_action(ccb& data, multi_ex& examples)
+{
+  int32_t action_index = -1;
+  for (size_t i = 0; i < examples.size(); i++)
+  {
+    CB::label& ld = examples[i]->l.cb;
```
Comment (jackgerrits) [inline vowpalwabbit/conditional_contextual_bandit.cc:220]:
```suggestion
    const CB::label& ld = examples[i]->l.cb;
```

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 5]
```diff
@@ -211,6 +212,26 @@ void save_action_scores(ccb& data, decision_scores_t& decision_scores)
   data.shared->pred.a_s.clear();
 }
 
+void remove_labeled_action(ccb& data, multi_ex& examples)
```
Comment (jackgerrits) [inline vowpalwabbit/conditional_contextual_bandit.cc:215]:
```suggestion
void remove_labeled_action(ccb& data, const multi_ex& examples)
```

**Review by jackgerrits — APPROVED** [review]

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 6]
```diff
@@ -211,6 +212,26 @@ void save_action_scores(ccb& data, decision_scores_t& decision_scores)
   data.shared->pred.a_s.clear();
 }
 
+void remove_labeled_action(ccb& data, const multi_ex& examples)
```
Comment (olgavrou) [inline vowpalwabbit/conditional_contextual_bandit.cc:215]:
```suggestion
void add_chosen_action_to_exclude_list(ccb& data, const multi_ex& examples)
```

or something similar, remove implies that something is actually removed, i.e. that this method has side effects

**Review by olgavrou — APPROVED** [review]
approve with a suggestion

**Pull request merged**